vyper/semantics/analysis/local.py

```diff
@@ -759,33 +720,32 @@ def _analyse_range_call(node: vy_ast.Call) -> list[VyperType]:
     validate_call_args(node, (1, 2), kwargs=["bound"])
     kwargs = {s.arg: s.value for s in node.keywords or []}
     start, end = (vy_ast.Int(value=0), node.args[0]) if len(node.args) == 1 else node.args
-    start, end = [i.get_folded_value() if i.has_folded_value else i for i in (start, end)]
-
-    all_args = (start, end, *kwargs.values())
-    for arg1 in all_args:
-        validate_expected_type(arg1, IntegerT.any())
+    folded_start, folded_end = [
+        i.get_folded_value() if i.has_folded_value else i for i in (start, end)
+    ]
 
-    type_list = get_common_types(*all_args)
-    if not type_list:
-        raise TypeMismatch("Iterator values are of different types", node)
+    all_args_unfolded = (start, end, *kwargs.values())
+    all_args_folded = (folded_start, folded_end, *kwargs.values())
+    for unfolded_arg, folded_arg in zip(all_args_unfolded, all_args_folded):
+        try:
+            validate_expected_type(folded_arg, iter_type)
+        except VyperException as e:
+            raise InvalidType(str(e), unfolded_arg)
```

Comment:
i don't think we need to be so intentional here, finding the proper location should be the responsibility of some scaffolding somewhere

Reply:
actually we don't need to run `validate_expected_type` since everything is already visited by the expr visitor

Reply:
i fixed this in https://github.com/vyperlang/vyper/pull/3596/commits/77d9bd370c48ebe08ceae8bc0c1ebb831fc7f19b